code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +164,247 @@ public void updateAccount(Account account, String name, String kind, String exte
+        return lockManager.lock(new ProjectLock(project), new LockCallback<List<? extends ProjectMember>>() {
+            public List<? extends ProjectMember> doWithLock() {
+                List<? extends ProjectMember> previousMembers = getActiveProjectMembers(project.getId());
+                HashSet<Member> otherPreviosMembers = new HashSet<>();
+                for (ProjectMember member : previousMembers) {
+                    otherPreviosMembers.add(new Member(member));
+                }
+                HashSet<Member> update = new HashSet<>();
+                HashSet<Member> createIgnore = new HashSet<>();
+                HashSet<Member> deleteIgnore = new HashSet<>();
+                for (Member member : otherPreviosMembers) {
+                    for (Member memberNew : members) {
+                        if (memberNew.getExternalId().equalsIgnoreCase(member.getExternalId()) &&
+                                memberNew.getExternalIdType().equalsIgnoreCase(member.getExternalIdType())) {
+                            if (!memberNew.getRole().equalsIgnoreCase(member.getRole())) {
+                                update.add(memberNew);
+                                deleteIgnore.add(member);
+                            } else {
+                                createIgnore.add(memberNew);
+                                deleteIgnore.add(member);
+                            }
+                        }
+                    }
+                }
+                HashSet<Member> create = (HashSet) members.clone();
+                create.removeAll(update);
+                create.removeAll(createIgnore);
+                HashSet<Member> delete = (HashSet) otherPreviosMembers.clone();
+                delete.removeAll(deleteIgnore);
+
+                Condition allMembers = DSL.falseCondition();
+                for (Member member : delete) {
+                    allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                            .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                            .and(PROJECT_MEMBER.PROJECT_ID.eq(project.getId())));
+                }
+                List<? extends ProjectMember> toDelete = create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch();
+                for (ProjectMember member : toDelete) {
+                    objectProcessManager.executeStandardProcess(StandardProcess.DEACTIVATE, member, null);
+                    objectProcessManager.executeStandardProcess(StandardProcess.REMOVE, member, null);
+                }
+                List<ProjectMember> newMembers = new ArrayList<>();
+                for (Member member : create) {
+                    newMembers.add(createProjectMember(project, member));
+                }
+                for (Member member : update) {
+                    newMembers.add(updateMemberRole(project, member));
```

Comment:
We could also remove the objectManager.reload(gotMember) at the end of updateMemberRole() and just opt for a single SELECT statement after this for loop.  Otherwise, it's 2 SELECTs and 1 UPDATE for every member we're updating.